pkg/ebpf/tracer.go

```diff
@@ -766,38 +772,48 @@ func (t *Tracer) DebugNetworkMaps() (*network.Connections, error) {
 // closed ports will be returned.
 // the map will be one of port_bindings  or udp_port_bindings, and the mapping will be one of tracer#portMapping
 // tracer#udpPortMapping respectively.
-func (t *Tracer) populatePortMapping(mp *ebpf.Map, mapping *network.PortMapping) ([]uint16, error) {
-	var key, emptyKey uint16
+func (t *Tracer) populatePortMapping(mp *ebpf.Map, mapping *network.PortMapping) (closed []portBindingTuple, closedNetNs []uint64, err error) {
+	var key, emptyKey portBindingTuple
 	var state uint8
 
-	closedPortBindings := make([]uint16, 0)
-
+	inos := make(map[int]uint64)
 	entries := mp.IterateFrom(unsafe.Pointer(&emptyKey))
 	for entries.Next(unsafe.Pointer(&key), unsafe.Pointer(&state)) {
-		port := key
+		pid := int(key.pid)
+		port := uint16(key.port)
+		var ino uint64
+		var ok bool
+		if ino, ok = inos[pid]; !ok {
+			var err error
+			ino, err = util.GetNetNsInoFromPid(t.config.ProcRoot, pid)
+			if err != nil {
+				log.Errorf("could not add port mapping (pid: %d port: %d), could not get net ns for pid %d: %s", pid, port, pid, err)
+				continue
+			}
 
-		mapping.AddMapping(port)
+			inos[pid] = ino
+		}
+
+		mapping.AddMappingWithNs(ino, port)
 
 		if isPortClosed(state) {
-			closedPortBindings = append(closedPortBindings, port)
+			closed = append(closed, key)
+			closedNetNs = append(closedNetNs, ino)
 		}
 	}
 
-	return closedPortBindings, nil
+	return closed, closedNetNs, nil
 }
 
 func (t *Tracer) determineConnectionDirection(conn *network.ConnectionStats) network.ConnectionDirection {
+	pm := t.portMapping
 	if conn.Type == network.UDP {
-		if t.udpPortMapping.IsListening(conn.SPort) {
-			return network.INCOMING
-		}
-		return network.OUTGOING
+		pm = t.udpPortMapping
 	}
-
-	if t.portMapping.IsListening(conn.SPort) {
+	if pm.IsListeningWithNs(uint64(conn.NetNS), conn.SPort) ||
+		(conn.IPTranslation != nil && pm.IsListeningWithNs(uint64(conn.NetNS), conn.IPTranslation.ReplDstPort)) {
```

Comment:
Is it correct to check both the NATed and non-NATed port ?

Reply:
If there is NAT translation there, only the nat check will succeed, but since we're using an OR it doesn't really matter.